app/components/slide_up_panel/slide_up_panel.js

```diff
@@ -47,189 +47,241 @@ export default class SlideUpPanel extends PureComponent {
-    };
-
-    startAnimation = (initialY, positionY, isGoingDown, initial = false) => {
-        const {allowStayMiddle, containerHeight, onRequestClose} = this.props;
-        const {finalPosition, initialPosition} = this.state;
-        const position = new Animated.Value(initial ? initialY : positionY);
-        let endPosition = (!isGoingDown && !initial ? finalPosition : positionY);
-
-        position.removeAllListeners();
-        if (isGoingDown) {
-            if (positionY <= this.state.initialPosition && allowStayMiddle) {
-                endPosition = initialPosition;
+    onHandlerStateChange = ({nativeEvent}) => {
+        if (nativeEvent.oldState === GestureState.ACTIVE) {
+            const {translationY, velocityY} = nativeEvent;
+            const {allowStayMiddle, containerHeight} = this.props;
+            const {lastSnap} = this.state;
+            const isGoingDown = translationY > 0;
+            const translation = translationY - this.lastScrollYValue;
+
+            const endOffsetY = lastSnap + translation;
+            let destSnapPoint = this.snapPoints[0];
+
+            if (Math.abs(translationY) < 50) {
+                destSnapPoint = lastSnap;
+            } else if (isGoingDown && !allowStayMiddle) {
+                destSnapPoint = containerHeight;
+            } else if (isGoingDown) {
+                destSnapPoint = this.snapPoints.find((s) => s >= endOffsetY);
             } else {
-                endPosition = containerHeight;
+                destSnapPoint = this.snapPoints.find((s) => s <= endOffsetY);
             }
-        }
-
-        Animated.timing(position, {
-            toValue: endPosition,
-            duration: initial ? 200 : 100,
-            useNativeDriver: true,
-        }).start(() => {
-            if (this.viewRef && this.backdrop) {
-                this.setState({endPosition});
-                this.backdrop.setNativeProps({pointerEvents: 'box-only'});
-                this.isDragging = false;
 
-                if (endPosition === containerHeight) {
-                    onRequestClose();
+            if (destSnapPoint) {
+                this.translateYOffset.extractOffset();
+                this.translateYOffset.setValue(translationY);
+                this.translateYOffset.flattenOffset();
+                this.dragY.setValue(0);
+
+                if (destSnapPoint === containerHeight) {
+                    this.closeWithAnimation();
+                } else {
+                    Animated.spring(this.translateYOffset, {
+                        velocity: velocityY,
+                        tension: 68,
+                        friction: 12,
+                        toValue: destSnapPoint,
+                        useNativeDriver: true,
+                    }).start(() => {
+                        this.setState({lastSnap: destSnapPoint});
+                    });
                 }
+            } else {
+                Animated.spring(this.translateYOffset, {
```

Comment:
When could we hit this case? Maybe when overscrolling?

Reply:
this is only a safety net in case the endOffset does not return a snapping position